src/pudl/etl/__init__.py

```diff
@@ -95,6 +103,61 @@
     ),
 )
 
+
+def asset_check_from_schema(
+    asset_key: AssetKey,
+    package: pudl.metadata.classes.Package,
+) -> AssetChecksDefinition | None:
+    """Create a dagster asset check based on the resource schema, if defined."""
+    resource_id = asset_key.to_user_string()
+    try:
+        resource = package.get_resource(resource_id)
+    except ValueError:
+        return None
+    pandera_schema = resource.schema.to_pandera()
+
+    @asset_check(asset=asset_key)
```

Comment:
So this is the line that says "Asset X gets asset check Y?" I definitely didn't catch that the first time through.

This is efficient / concise but feels a little opaque. I think anything we can do to make the definition of new asset checks / validations intuitive and low-overhead will help us add lots of them incrementally over time, which will be good for data quality.  What are the different categories / modes of definition that we're going to enable?

- Autogenerated checks based on the `TableSchema` / SQL checks.
- Re-usable, parameterized classes of data validations that are specified in the Resource definition, and so also end up getting injected here (like the ones you created from our old data validation for the boiler fuel table, but with standalone check classes/parameters)
- Additional more bespoke per-asset checks which are defined close to the asset definition and added with an explicit `@asset_check` decorator on the asset?

Are there any others?

Reply:
I think that's right! I'm inclined to go one of two ways:

## Super Simple
* **only** autogenerate type checks from schema, and undo the `df_checks` and `field_checks` machinery in this PR
* all other type checks go into colocated `@asset_check` definitions, and clean up the `pudl.validations` module to make those re-usable data validations simple to add

Benefits being: less weird asset check meta-machinery to maintain. Costs being that checks such as "the 5th percentile of this column, weighted by this other column, should be less than X value" might be nice to see *in the schema itself*.

## Slightly more complicated
* keep the `df/field_checks` machinery, clean up `pudl.validations`
* complicated checks still live in colocated `@asset_check`s.

Benefits: for the most common cases, all the checks will live in the `RESOURCE_METADATA`
Costs: we have a bit more machinery to maintain.

I guess in either case we need to clean up pudl.validations a bit more. I think I slightly lean towards "super simple" because that's less custom logic for us to maintain.

Reply:
I'm not sure which of these arrangements sounds better to me. It would be nice if there were *some* centralized place where you can easily see all of the asset checks that are being applied to an asset. But maybe that's already available within the Dagster machinery somewhere? If so, then using that built-in functionality and avoiding building and maintaining something in parallel seems like a good idea.

One reason we haven't developed more and better validations, and haven't applied the same kinds of validations we do have for our older tables to the (hundreds of!) new ones we've created is that the current validation specification system is hard to work with, and with the change in caching (all data read from the DB every time, rather than being cached in memory inside `PudlTabl`) the validations are now very slow to run.

In converting the commonly applied validations that we've currently specified, it seems like we should be setting up a pattern that's easy for us to incrementally expand as we apply these kinds of validations to new tables, and also come up with new things to validate.  If it's low computational overhead (because we're applying them when the data is already in memory) and low human overhead (because it's easy to specify a new validation using the existing classes as templates) then we'll hopefully see a proliferation of validations, and having a clean pattern to accommodate them will help keep it from turning into spaghetti.

One problem that's going to come up here is that we'll expect different outcomes depending on whether we're running the full or fast ETL. E.g. looking at gas prices or coal heat content just for 2022 will yield a different expected distribution than looking at the last 20 years of data, so we'll either need to change the expectations to be more lenient and line up with all years / any individual year of data, or we'll need to be able to specify different asset check expectations for the fast vs. full ETL.

Translating all of the existing validations directly is also probably not ideal. I didn't really design them very well.  E.g. the min/max rows checks were originally meant to check that the number of rows were within a particular range, but now we really just check for an exact expected row count. I think there's also some unnecessary redundancy in the weighted distribution checks and some of them can probably just be discarded (I'm thinking of the `vs_self()` checks here). I think it may also be that we can re-use the same cumulative sums for calculating all of the weighted quantiles we're interested in, rather than recalculating them for every check, which could cut the time for doing these checks down by ~75%.

So... which of these many possible threads do we want to pull on right now? And which do we want to defer until we hopefully get some additional data validation / CI robustness support via POSE?

To me it feels like establishing an easy to use, extensible pattern for defining new data validations that makes use of Pandera and the asset checks, and starting to use that pattern for the new data we're integrating is a high priority (stop digging the janky data validation hole any deeper), but converting all of the existing jank into that new form might be lower priority, except insofar as those validations inform the needs we have for the new system.

Reply:
> It would be nice if there were _some_ centralized place where you can easily see all of the asset checks that are being applied to an asset. But maybe that's already available within the Dagster machinery somewhere?

Yep! You can look in the `checks` tab in the asset catalog.

> To me it feels like establishing an easy to use, extensible pattern for defining new data validations that makes use of Pandera and the asset checks, and starting to use that pattern for the new data we're integrating is a high priority (stop digging the janky data validation hole any deeper), but converting all of the existing jank into that new form might be lower priority, except insofar as those validations inform the needs we have for the new system.

Agreed. I think that looks like utility methods in `pudl.validate` that make it easy to create asset checks for:
* row count
* weighted distribution - tail size, but also "peaks at X, Y, Z"
* no missing years of data, based on the dataset settings


I think the todo list then is:

* [ ] rip out the `df_checks` and `field_checks` machinery (though, keeping the frictionless compatibility stuff with uniqueness/min-max)
* [ ] merge this PR
* [ ] add these utility methods as a separate PR & use them to define a few checks

Reply:
## Decisions
* Agree let's keep all of the Frictionless / schema / datapackage compatible validations specified in the schema / Resource area, and move the non-schema checks to live near the assets they apply to.
* With that, the threshold for getting this PR merged is just that all of the schema validation checks are being run, and we make a new PR to set up the classes and example structures for the existing re-usable non-schema checks.
* An additional reusable non-schema check that we do on many tables and should probably do on many more is verify that there are no entirely NULL columns. There's a `pudl.validate.no_null_cols()` function that does this I think, which should be added to the list of utility methods to parameterize.

## Thoughts & Questions
* Checking for the exact expected row counts should be extremely fast if we just use values from `df.shape`
* Agree no missing years (or maybe more precisely, no missing partitions?) is important. I think we've had some issues with this recently in the `eia860m`, where the settings and archive were updated to include the new months of data, but the `working_partitions` associated with the data source were not updated, so none of the new data actually made it through the ETL.
* There are also some cases in which we set an allowable missignness threshold for rows, via `pudl.validate.no_null_rows()` but I think that's only applied in the case of the MCOE outputs, so maybe we don't need to parameterize it right now.
* Should/Can the schema-based asset checks ensure uniqueness across the set of PK columns when there is a PK defined? It would be great to identify that kind of problem as soon as it appears rather than only later in the DB, and getting those constraints into the asset checks would make them independent of the storage medium, so if we were to only output Parquet, they'd still be applied.
* The ability to define asset checks that rely on more than one asset as inputs makes me wonder about creating a generic, storage-independent referential integrity check -- if we can check that FK relationships are satisfied even when we're writing out to Parquet or BigQuery that would be great.

Reply:
Seems like we can validate joint uniqueness by taking the `primary_key` and shoving it into the Pandera schema's `unique` constructor param:

https://pandera.readthedocs.io/en/stable/dataframe_schemas.html#validating-the-joint-uniqueness-of-columns

Reply:
Oh, I actually already do that in this PR. No changes necessary!